File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// Helper function to reverse a linked list
func reverseList(head *ListNode) *ListNode {
    var prev *ListNode
    current := head

    for current != nil {
        next := current.Next
        current.Next = prev
        prev = current
        current = next
    }
    return prev
}

// Function to add two numbers represented by linked lists
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    // Reverse both lists
    l1 = reverseList(l1)
    l2 = reverseList(l2)

    dummy := &ListNode{}
    current := dummy
    carry := 0

    // Traverse both reversed linked lists
    for l1 != nil || l2 != nil || carry != 0 {
        var x, y int
        if l1 != nil {
            x = l1.Val
            l1 = l1.Next
        }
        if l2 != nil {
            y = l2.Val
            l2 = l2.Next
        }

        // Calculate the sum and carry
        sum := x + y + carry
        carry = sum / 10

        // Create a new node with the sum value (mod 10) and add to the result list
        current.Next = &ListNode{Val: sum % 10}
        current = current.Next
    }

    // Reverse the result list to get the final answer in correct order
    return reverseList(dummy.Next)
}

File: 0445-add-two-numbers-ii/0445-add-two-numbers-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        def reverse(head):
            prev=None
            curr=head
            while curr:
                nxt=curr.next
                curr.next=prev
                prev=curr
                curr=nxt
            return prev
        l1=reverse(l1)
        l2=reverse(l2)
        carry=0
        dummy=ListNode()
        temp=dummy
        while l1 and l2:
            s=l1.val+l2.val+carry
            temp.next=ListNode(val=s%10)
            carry=s//10
            temp=temp.next
            l1=l1.next
            l2=l2.next
        l=l1 if l1 else l2
        while l:
            s=l.val+carry
            temp.next=ListNode(val=s%10)
            carry=s//10
            temp=temp.next
            l=l.next
        if carry>0: 
            temp.next=ListNode(val=carry)
            temp=temp.next
        dummy=dummy.next
        return reverse(dummy)